pkg/compact/downsample/downsample_test.go

```diff
@@ -162,185 +162,389 @@ func TestExpandChunkIterator(t *testing.T) {
-		if len(d.inRaw) > 0 && len(d.inAggr) > 0 {
-			t.Fatalf("test must not have raw and aggregate input data at once")
-		}
-		ser := &series{lset: d.lset}
+			dir, err := ioutil.TempDir("", "downsample-raw")
+			testutil.Ok(t, err)
+			defer func() { testutil.Ok(t, os.RemoveAll(dir)) }()
 
-		if len(d.inRaw) > 0 {
-			chk := chunkenc.NewXORChunk()
-			app, _ := chk.Appender()
+			// Ideally we would use tsdb.HeadBlock here for less dependency on our own code. However,
+			// it cannot accept the counter signal sample with the same timestamp as the previous sample.
+			mb := newMemBlock()
+			ser := chunksToSeriesIteratable(t, tcase.inRaw, tcase.inAggr)
+			mb.addSeries(ser)
 
-			for _, s := range d.inRaw {
-				app.Append(s.t, s.v)
+			fakeMeta := &metadata.Meta{}
+			if len(tcase.inAggr) > 0 {
+				fakeMeta.Thanos.Downsample.Resolution = tcase.resolution - 1
 			}
-			ser.chunks = append(ser.chunks, chunks.Meta{
-				MinTime: d.inRaw[0].t,
-				MaxTime: d.inRaw[len(d.inRaw)-1].t,
-				Chunk:   chk,
-			})
-		} else {
-			ser.chunks = append(ser.chunks, encodeTestAggrSeries(d.inAggr))
-		}
-		mb.addSeries(ser)
-	}
 
-	id, err := Downsample(log.NewNopLogger(), meta, mb, dir, resolution)
-	testutil.Ok(t, err)
+			id, err := Downsample(logger, fakeMeta, mb, dir, tcase.resolution)
+			if tcase.expectedDownsamplingErr != nil {
+				testutil.NotOk(t, err)
+				testutil.Equals(t, tcase.expectedDownsamplingErr(ser.chunks).Error(), err.Error())
+				return
+			}
+			testutil.Ok(t, err)
 
-	_, err = metadata.Read(filepath.Join(dir, id.String()))
-	testutil.Ok(t, err)
+			_, err = metadata.Read(filepath.Join(dir, id.String()))
+			testutil.Ok(t, err)
 
-	exp := map[uint64]map[AggrType][]sample{}
-	got := map[uint64]map[AggrType][]sample{}
+			indexr, err := index.NewFileReader(filepath.Join(dir, id.String(), block.IndexFilename))
+			testutil.Ok(t, err)
+			defer func() { testutil.Ok(t, indexr.Close()) }()
 
-	for _, d := range data {
-		exp[d.lset.Hash()] = d.output
-	}
-	indexr, err := index.NewFileReader(filepath.Join(dir, id.String(), block.IndexFilename))
-	testutil.Ok(t, err)
-	defer func() { testutil.Ok(t, indexr.Close()) }()
+			chunkr, err := chunks.NewDirReader(filepath.Join(dir, id.String(), block.ChunksDirname), NewPool())
+			testutil.Ok(t, err)
+			defer func() { testutil.Ok(t, chunkr.Close()) }()
 
-	chunkr, err := chunks.NewDirReader(filepath.Join(dir, id.String(), block.ChunksDirname), NewPool())
-	testutil.Ok(t, err)
-	defer func() { testutil.Ok(t, chunkr.Close()) }()
+			pall, err := indexr.Postings(index.AllPostingsKey())
+			testutil.Ok(t, err)
 
-	pall, err := indexr.Postings(index.AllPostingsKey())
-	testutil.Ok(t, err)
+			var series []uint64
+			for pall.Next() {
+				series = append(series, pall.At())
+			}
+			testutil.Ok(t, pall.Err())
+			testutil.Equals(t, 1, len(series))
 
-	for pall.Next() {
-		id := pall.At()
+			var lset labels.Labels
+			var chks []chunks.Meta
+			testutil.Ok(t, indexr.Series(series[0], &lset, &chks))
```

Comment:
Very minor suggestion: we could be a bit paranoic here and check if `lset` has the right labels.

Reply:
:+1:  Good point!